chore(docs): Add section regarding Subscriptions Services data sources (#338)
mozilla/ecosystem-platform: docs/relying-parties/reference/metrics-for-relying-parties.md

sidebar_label: Metrics
---

Last updated: `March 15th, 2023`

## Subscriptions Services Data Sources
Subscription Platform provides product metrics by obtaining subscriptions data from the following sources: Stripe, mobile app store providers, and FxA logs.

### Stripe

> The bulk of Subscriptions Services data comes from Stripe and is synced to BigQuery using FiveTran and its Stripe connector. This data reflects the current subscription state of all active subscriptions paid for via Stripe or PayPal. It includes information, such as the customer's start and end dates of a subscription, etc.

Author: Help: what other information is included?

Reviewer: The following paragraph lists some other customer information that's included.

In addition to subscriptions and customers we also have products, plans, invoices, cards, charges, discounts, coupons, and promotion codes.

> Other information includes customers' postal code, state, and country for a subscription, which helps resolve issues for customers across a particular dimension (e.g., country, payment provider, etc.). As the data intentionally excludes any personal identifiable information of customers (e.g., name, street address, phone number, email address), it makes it challenging to use this data to help on an individual customer level. Other uses for this data is for retrieving time counts (e.g., total number of active subscriptions).

Author: Help: the [ticket](https://mozilla-hub.atlassian.net/browse/FXA-6943) states that Stripe data is "useful for all time counts of things like total number of (active) subscriptions". What does "time counts" mean, what else does that include?

Reviewer: I think that means _all-time_ counts (i.e. total counts).

> While FiveTran does not support capturing subscription change events, it preserves historical data within `subscription_history` by comparing the current state of the subscription with the most recent record in the history table and adding a new record if they vary.

Author: Thanks for the clarification, Sean! I have opted to remove `Fivetran does not support capturing subscription change events` and just keep the rest of the statement.

> Fivetran preserves historical data within `subscription_history` by comparing the current state of the subscription with the most recent record in the history table and adding a new record if they vary.

Reviewer 2: IIUC this is limited in terms of what history we can get and/or is difficult to identify what changed between one version of a Stripe subscription object and a later version, which is why we will also rely on a Stripe Firestore sync for historical subscription data. @sean-rose , can you confirm this? It'd be nice to discuss these limitations briefly if so and mention that as of February 2023, we also sync SubPlat's Stripe Firestore Customer and Subscriptions collections (see https://mozilla-hub.atlassian.net/browse/SVCSE-1080).

Reviewer: > IIUC this is limited in terms of what history we can get and/or is difficult to identify what changed between one version of a Stripe subscription object and a later version, which is why we will also rely on a Stripe Firestore sync for historical subscription data. @sean-rose , can you confirm this?

The main issue with the Fivetran Stripe sync is it only records historical state/changes for the top-level subscription fields and not the associated subscription items which indicate the subscribed plan, so we have to rely on the subscription metadata fields `previous_plan_id` and `plan_change_date`.  However, those only record the most recent plan change, so if there are any problems with the Fivetran sync (like [happened last month](https://mozilla-hub.atlassian.net/browse/DENG-754)) then we can miss/lose history.

So yes, the current plan is to make use of the Stripe Firestore changelog data, but before actually doing that we'll need to verify that the Stripe Firestore collection is reliable (e.g. this week I found some issues with Firestore missing some subscriptions), and since the Stripe Firestore collection doesn't contain all historical subscriptions/changes we'll still need to use the Fivetran data for some subscriptions.

> It'd be nice to discuss these limitations briefly if so and mention that as of February 2023, we also sync SubPlat's Stripe Firestore Customer and Subscriptions collections (see https://mozilla-hub.atlassian.net/browse/SVCSE-1080).

It'd also be worth noting when the Stripe Firestore collections themself were created.

Reviewer 2: > So yes, the current plan is to make use of the Stripe Firestore changelog data, but before actually doing that we'll need to verify that the Stripe Firestore collection is reliable (e.g. this week I found some issues with Firestore missing some subscriptions), and since the Stripe Firestore collection doesn't contain all historical subscriptions/changes we'll still need to use the Fivetran data for some subscriptions.

Yep we've been discussing this in [Slack](https://mozilla.slack.com/archives/CG0AJ6E77/p1679605292654949). As I mentioned there, these aren't bugs in our implementation, but rather represent a new use case that differs from our original intent/design.

Reviewer 2: > It might also be worth noting when the Stripe Firestore collections themselves were created.

This is October 2021 per [FXA-4023](https://mozilla-hub.atlassian.net/browse/FXA-4023).

### Mobile app store providers
Subscriptions Services data is also obtained from two mobile app store providers: Apple and Google. This includes date of billing and expired dates, subscription change event, etc. Apple provides limited data - it does not provide the country, price, or currency of subscriptions.

> Subscription Platform handles getting Apple subscriptions from the Apple API and stores the latest state of active subscriptions in Firestore. Prior to December 2022, VPN Apple subscriptions data was managed by VPN's Guardian service and stored in the Guardian database. As Firestore doesn't contain the full historical state of Apple subscriptions, Guardian data is still used for historical purposes.

Reviewer 2: Nit: s/Apple API/App Store Server API

I'm wondering if we can shorten the VPN note about using the Guardian DB in addition to SubPlat's Firestore DB. Ideally, we want statements here to be generally true/applicable to all RPs with minimal RP-specific asides. Granted, Sean's talk had a heavy emphasis on VPN-specific details, because they were the first major SubPlat product to go through the ringer here. I would imagine other product teams won't be in this situation of having older subscription data.

Maybe something like: As of December 2022, SubPlat supports [Apple IAP](link to our docs on Apple IAP] subscriptions, with current and historical subscription state from that point forward stored in our Firestore DB and synced in BigQuery.

> Similar to Apple, Subscription Platform also handles the current and historical state of Google subscriptions and it is synced to Firestore as well.

Reviewer 2: Getting a bit technical here, but only current state for IAP subscriptions (Apple and Google) are stored in Firestore. We have a BigQuery sync that syncs these two collections and a separate table for each that syncs the changelogs, which is where we can get historical state from.

Reviewer: Getting more technical:  the changelog for each collection is what gets synced to a BigQuery table, and there's a BigQuery view for each that represents the current state of the collection calculated based on the latest changelog for each document.

Since the historical state/changes are only recorded by the BigQuery sync itself, we only have records of the changes that occurred after the BigQuery syncs were set up, so it might be worth noting when those syncs were set up:
- Google IAP sync started 2021-10-18.
- Apple IAP sync started 2022-12-15.

> The general use of this data is to understand what the subscription state was at the time an event occurred (e.g., when Stripe products and plans change) as well as sequences of events.

Reviewer: I'd suggest removing the comment "(e.g., when Stripe products and plans change)" as this section is about Apple/Google subscriptions.  The similar comment in [FXA-6943](https://mozilla-hub.atlassian.net/browse/FXA-6943) was related to the new SubPlat Stripe Firestore collection syncs which were recently set up, and while those Stripe Firestore syncs may eventually be used in addition to or instead of the Fivetran Stripe sync, they aren't currently used for this purpose.
```suggestion
The general use of this data is to understand what the subscription state was over time, as well as sequences of events.
```

### FxA logs

> FxA logs include event data that can help provide insights regarding the customers' journey through the FxA sign-in/sign-up process. It can also help identify potential drivers (e.g., call-to-action, medium, paid search term, etc.) that directed customers to a page for acquisition and attribution purposes, if query parameters are set up properly (see below for more information).

Reviewer 2: We have all kinds of user events here. Maybe also note account and subscription management events? E.g. we'll be implementing a subscription ended event as part of our current metrics epic: https://mozilla-hub.atlassian.net/browse/FXA-6955.

Some of the customer information includes the customer's postal code, state, and country for a subscription, which helps resolve issues for customers across a particular dimension (e.g., country, payment provider, etc.). However, as it intentionally excludes any personal identifiable information of customers (e.g., name, street address, phone number, email address), it can be challenging to use to help on an individual customer level. Other uses for this data is for retrieving all-time counts (e.g., total number of active subscriptions).

Fivetran preserves historical data within `subscription_history` by comparing the current state of the subscription with the most recent record in the history table and adding a new record if they vary. However, there are limitations with the Fivetran Stripe sync, as it only records historical state/changes for top-level subscription fields and not the associated subscription items which indicate the subscribed plan. As we have to rely on the subscription metadata fields `previous_plan_id` and `plan_change_date`, which only records the most recent plan change, we can potentially miss/lose history if there are any problems with the Fivetran sync.

> As of February 2023, Subscription Platform also syncs Stripe Firestore Customer and Subscriptions collections.

Reviewer 2: Can we add: "in BigQuery including the latest data and changelogs" and mention that the Stripe Firestore DB itself was set up in October 2021 as Sean suggested elsewhere?

### Mobile app store providers
Subscriptions Services data is also obtained from two mobile app store providers: Apple and Google. This includes date of billing and expired dates, subscription change event, etc. Apple provides limited data - it does not provide the country, price, or currency of subscriptions.

> As of December 2022, Subscription Platform supports [Apple IAP][apple-iap-integration] subscriptions, and stores both current and historical subscription state from that point forward in the Firestore database and synced in BigQuery.

Reviewer 2: s/current and historical/current? You have a paragraph further down that covers historical state.
